kubemarine/kubernetes/__init__.py

```diff
@@ -926,21 +957,37 @@ def upgrade_workers(version, upgrade_group, cluster, kubeadm_file, minor_version
         exclude_node_from_upgrade_list(first_control_plane, node['name'])
 
 
-def prepare_drain_command(node, version: str, globals, disable_eviction: bool, nodes,
+def prepare_drain_command(cluster: KubernetesCluster, node_name: str,
+                          *,
+                          disable_eviction=False,
                           drain_timeout: int = None, grace_period: int = None):
-    drain_globals = globals['nodes']['drain']
+    drain_globals = static.GLOBALS['nodes']['drain']
     if drain_timeout is None:
-        drain_timeout = recalculate_proper_timeout(nodes, drain_globals['timeout'])
+        drain_timeout = recalculate_proper_timeout(cluster.nodes, drain_globals['timeout'])
+
     if grace_period is None:
         grace_period = drain_globals['grace_period']
-    drain_cmd = f"kubectl drain {node['name']} --force --ignore-daemonsets --delete-emptydir-data " \
+
+    drain_cmd = f"kubectl drain {node_name} --force --ignore-daemonsets --delete-emptydir-data " \
                 f"--timeout={drain_timeout}s --grace-period={grace_period}"
-    if version and disable_eviction:
+    if disable_eviction:
         drain_cmd += " --disable-eviction=true"
     return drain_cmd
 
 
-def upgrade_cri_if_required(group):
+def delete_system_pods(cluster: KubernetesCluster, control_plain: NodeGroup):
```

Comment:
Why do we need to delete apiserver/etcd/scheduler/controller-manager pods? As I understand deleting these pods will not cause actual containers restart and the pods will be immediately re-created, because they are static. So as for me it seems redundant to delete static pods

Reply:
It is just done this way in `migrate_cri` procedure #13 and I copied the approach to `migrate_kubemarine`. I can remove this method and its invocations from both procedures, if everyone and especially @alexarefev agrees.

Reply:
I think you are mixing up different cases. The system pods deletion is necessary for further steps inside the `migrate_cri` procedure. It just refresh the starting time for system pods after that system pods are killed by `docker rm`. Maybe it's not the best solution for `migrate_cri` but I'm not sure that we need extend this practice on `migrate_kubemarine` procedure.

Reply:
> The system pods deletion is necessary for further steps inside the migrate_cri procedure. It just refresh the starting time for system pods after that system pods are killed by docker rm.

But why system pods deletion ([migrate_cri.py#L165](https://github.com/Netcracker/KubeMarine/blob/0.17.0/kubemarine/procedures/migrate_cri.py#L165)) deletion happens before docker rm ([migrate_cri.py#L187](https://github.com/Netcracker/KubeMarine/blob/0.17.0/kubemarine/procedures/migrate_cri.py#L187)) then?

Reply:
> > The system pods deletion is necessary for further steps inside the migrate_cri procedure. It just refresh the starting time for system pods after that system pods are killed by docker rm.
> 
> But why system pods deletion ([migrate_cri.py#L165](https://github.com/Netcracker/KubeMarine/blob/0.17.0/kubemarine/procedures/migrate_cri.py#L165)) happens **before** docker rm ([migrate_cri.py#L187](https://github.com/Netcracker/KubeMarine/blob/0.17.0/kubemarine/procedures/migrate_cri.py#L187)) then?

Because `docker rm` is applicable after the `kubelet` service stop

Reply:
> refresh the starting time for system pods after that system pods are killed by docker rm.

Ok, so I conclude that this statement is not true, and it does not do what was intended.

Reply:
Deleted the `delete_system_pods` invocation from both `migrate_cri` and `migrate_kubemarine`. Checked migration and CRI upgrade on (2 balancers, 2 control planes, 1 worker) scheme, it was successful.

Reply:
> > refresh the starting time for system pods after that system pods are killed by docker rm.
> 
> Ok, so I conclude that this statement is not true, and it does not do what was intended.

I don't understand which statement you've mentioned

Reply:
@alexarefev
> refresh the starting time for system pods after that system pods are killed by docker rm.

Reply:
> @alexarefev
> 
> > refresh the starting time for system pods after that system pods are killed by docker rm.

Ok, How do you suggest to solve the [issue](https://github.com/Netcracker/KubeMarine/pull/13) ?

Reply:
PR13 does not solve that issue.

Reply:
> PR13 does not solve the issue it describes.

Ok, what the solution should look like in your opinion?

Reply:
It is a matter of separate activity and PR.

Reply:
Ok, I got it. It's a top secret.